plugins/flatpak/gs-flatpak.c

```diff
@@ -3396,6 +3411,167 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	g_clear_object (&helper->cancellable);
+	g_clear_error (&helper->error);
+	g_free (helper);
+}
+
+static GsFlatpakCopyProcessHelper *
+gs_flatpak_copy_process_helper_new (GsFlatpak *self,
+				    GsApp *app,
+				    GCancellable *cancellable,
+				    gulong cancelled_id)
+{
+	GsFlatpakCopyProcessHelper *helper = g_new0 (GsFlatpakCopyProcessHelper,
+						     1);
+	helper->self = self;
+	helper->app = g_object_ref (app);
+	helper->cancellable = g_object_ref (cancellable);
+	helper->cancelled_id = cancelled_id;
+	helper->finished = FALSE;
+	helper->error = NULL;
+	return helper;
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GsFlatpakCopyProcessHelper, gs_flatpak_copy_process_helper_free)
+
+static void
+copy_process_watch_cb (GPid pid, gint status, gpointer user_data)
+{
+	GsFlatpakCopyProcessHelper *helper = user_data;
+	g_autoptr(GError) error = NULL;
+
+	if (!g_cancellable_is_cancelled (helper->cancellable) && status != 0)
+		g_set_error (&helper->error,
+			     GS_PLUGIN_ERROR,
+			     GS_PLUGIN_ERROR_FAILED,
+			     "failed to copy flatpak app to removable "
+			     "media: `flatpak create-usb` failed with "
+			     "status %d", status);
+
+	g_cancellable_disconnect (helper->cancellable, helper->cancelled_id);
+	g_spawn_close_pid (pid);
+
+	helper->finished = TRUE;
+
+	/* once the copy terminates (successfully or not), set plugin status to
+	 * update UI accordingly */
+	gs_plugin_status_update (helper->self->plugin, helper->app,
+				 GS_PLUGIN_STATUS_FINISHED);
+
+	/* update metadata to ensure the USB category includes this new app and
+	 * we can immediately download an app, copy to USB, delete from
+	 * computer, and install from USB without leaving the app page */
+	gs_flatpak_refresh (helper->self, 0, GS_PLUGIN_REFRESH_FLAGS_METADATA,
+			    NULL, &error);
+	if (error != NULL)
+		g_warning ("failed to refresh flatpak metadata after copying "
+			   "app to USB: %s", error->message);
+}
+
+static void
+app_copy_cancelled_cb (GCancellable *cancellable,
+		       gpointer user_data)
+{
+	GPid pid = GPOINTER_TO_INT (user_data);
+
+	/* terminate the process which is copying the app */
+	kill (pid, SIGTERM);
+}
+
+gboolean
+gs_flatpak_app_copy (GsFlatpak *self,
+		     GsApp *app,
+		     const gchar *copy_dest,
+		     GCancellable *cancellable,
+		     GError **error)
+{
+	/* this is used in an async function but we block here until that
+	 * returns so we won't auto-free while other threads depend on this */
+	g_autoptr (GsFlatpakCopyProcessHelper) helper = NULL;
+	const gchar *app_id = gs_flatpak_app_get_ref_name (app);
+	const gchar *app_arch = gs_flatpak_app_get_ref_arch (app);
+	const gchar *app_branch = gs_flatpak_app_get_ref_branch (app);
+	g_autofree gchar* app_ref = g_strdup_printf ("%s/%s/%s", app_id,
+						     app_arch, app_branch);
+	const gchar* app_collection_id = gs_flatpak_get_collection_id (self,
+								       app);
+
+	gboolean spawn_retval;
+	const gchar *argv[] = {"/usr/bin/flatpak", "create-usb", copy_dest,
```

Comment:
Isn't it bad to assume flatpak was installed into that location, especially if this will be upstreamed? GNOME Builder seems to not need to do that: https://gitlab.gnome.org/GNOME/gnome-builder/blob/master/src/plugins/flatpak/gbp-flatpak-runtime.c#L119

Reply:
Although in that case it's making sure `PATH` is set. So it might be better to use `which`.

Reply:
I still don't think this is the right approach (for example on NixOS binaries are installed into `/root/.nix-profile/bin/` rather than `/usr/bin/`) but there is precedent for it in the codebase (search for `DPKG_DEB_BINARY`) so I guess leave it and see what upstream says.